test/runtime/metrics.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package RuntimeTest
+
+import (
+	"strings"
+
+	"github.com/cilium/cilium/pkg/metrics"
+	. "github.com/cilium/cilium/test/ginkgo-ext"
+	"github.com/cilium/cilium/test/helpers"
+
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("RuntimeMetrics", func() {
+
+	var vm *helpers.SSHMeta
+
+	BeforeAll(func() {
+		vm = helpers.InitRuntimeHelper(helpers.Runtime, logger)
+		ExpectCiliumReady(vm)
+
+		areEndpointsReady := vm.WaitEndpointsReady()
+		Expect(areEndpointsReady).Should(BeTrue())
+	})
+
+	JustAfterEach(func() {
+		vm.ValidateNoErrorsInLogs(CurrentGinkgoTestDescription().Duration)
+	})
+
+	AfterFailed(func() {
+		vm.ReportFailed()
+	})
+
+	AfterAll(func() {
+		vm.CloseSSHClient()
+	})
+
+	Context("Metrics List CLI Test", func() {
+		defaultMetrics := metrics.DefaultMetrics()
```

Comment:
The test currently fails because `metrics.DefaultMetrics` returns all metrics, whether they're enabled or not. Maybe we can do the following instead:
```go
for _, metric := range viper.GetStringSlice(option.Metrics) {
	switch metric[0] {
	case '+':
		By("checking that metric '%s' is in 'cilium metric list' output", metric)
		...
``` 
That's the approach we use today to configure the Prometheus collectors:
https://github.com/cilium/cilium/blob/ad223e882df4dd1c427c8361942749fbb0dcfc8d/pkg/option/config.go#L2060-L2071

Reply:
Hi @pchaigno , thanks for the great comments! I'm actually quite confused around the enabling logics now, it seems that the code snippet you pointed to alters the default metrics list? In my local testing env, I didn't specify `metrics` in config, hence all default metrics should be registered? Is my understanding correct?

Reply:
but I agree that my original proposal was not sound :) maybe `metricsSlice` is what we should test against.

Reply:
> it seems that the code snippet you pointed to alters the default metrics list?

As I understand it, it retrieves the list of all metrics (`metrics.DefaultMetrics()`) then alters that list to keep only metrics that are actually enabled. It retrieves the list of enabled metrics through our configuration module `viper`. Neither the list of all metrics (those are constants defined in [`pkg/metrics/metrics.go`](https://github.com/cilium/cilium/blob/146eab12f6b919a15d66db1bf56a8a557d277821/pkg/metrics/metrics.go#L454)) nor the list of enabled metrics are modified here.

> maybe `c.MetricsConfig` is what we should test against.

[`config.MetricsConfig`](https://github.com/cilium/cilium/blob/146eab12f6b919a15d66db1bf56a8a557d277821/pkg/metrics/metrics.go#L399) would be great, but it won't give us the textual names of those metrics to match against the output. That's why I defaulted to using `viper.GetStringSlice(option.Metrics)`, from which `config.MetricsConfig` is constructed.

Reply:
Gotcha, I've modified the code to construct the same list as you suggested, PTAL and test. But I'm still quite confused about the enablement logic. My understanding is that after processing by `viper`, `metricsSlice` should have only enabled metrics, correct? I tried dumping it out, without specifying anything in config, it has all the default metrics in it, which seems to indicate those default metrics are enabled `by default`?

Reply:
It's possible that the tests disable some of the metrics... @raybejjani might know more.